spec/p2p/p2p-top-level.md

```diff
@@ -0,0 +1,310 @@
+
+- TMBC: Tendermint blockchain
+- SEQ: for sequential specifications
+- LIVE: liveness
+- SAFE: safety
+- FUNC: function
+- INV: invariant
+- A: assumption
+
+--->
+
+# Part I - A Tendermint node 
+
+TODO: 
+- perhaps we should survey here what we think are the expectations of the reactors from p2p? 
+    - consensus might need for liveness that neighborhood is stable
+    - mempool might just need that a transaction can reach each validator within a reasonable amount of time (this might be achievable if at no point in time the current graph is connected, but, e.g., something along the lines that the union of the graphs over some period in time is connected.)
+- What does p2p expect from the reactors? (don't falsely report bad nodes; this puts requirements on the reactors and perhaps/likely also on the application running on top of ABCI)
+
+## Context of this document
+
+> mention other components and or specifications that are relevant for this
+spec. Possible interactions, possible use cases, etc. 
+
+> should give the reader the understanding in what environment this component
+will be used. 
+
+TODO:
+- Reactor API
+- Network? How do we communicate with other nodes
+- Discuss that validators run special set-up, and manage their own neighborhood (hide behind sentry nodes).
+   - As a result: the distributed system is composed of
+       - (correct) nodes that follow the protocol described here
+       - (potentially) adversarial nodes whose behavior deviates to harm the system
+       - (correct) nodes that don't follow the protocol to shield themselves but behave in a "nice way"
+
+
+
+# Part II - Sequential Definition of the  Problem
+
+
+##  Informal Problem statement
+
+
+The p2p layer, specified here, manages the connections of a Tendermint node with other Tendermint nodes. It continuously provides a list of peers ensuring
+1. Connectivity. The overlay network induced by the local neighborhoods (defined by the lists of peers) is sufficiently connected so that the reactors can implement communication on top of it that is sufficient for their needs
+    > There is the design decision that the same overlay is used by all reactors. It seems that consensus has the strongest requirements regarding connectivity and this defines the required properties
+2. Stability. Typically, connections between correct peers should be stable
+    > Even if at every time *t* we satisfy Point 1, if the overlays at times *t* and *t+1* are totally different, it might be hard to implement decent communication on top of it. E.g., Consensus gossip requires a neighbor to know its neighbors *k* state so that it can send the message to *k* that help *k* to advance. If *k* is connected only one second per hour, this is not feasible.
+3. Openness. It is always the case that new nodes can be added to the system
+    > Assuming 1. and 2. holds, this means, there must always be nodes that are willing to add connections to new peers.
+
```

Comment:
Something about self-healing? (not sure, just thinking out loud)
Something like, "the network overlay can recover from important crippling network events (e.g. temporary partitions)"

Reply:
You are right. I wrote the properties 1.-3. having somewhat "good periods" in mind. The question as I see it now when I look at your comment is whether there is a relation between self healing and openness. But perhaps the connection is rather in the solution side. We need to think more ;-)